### ISSUE TITLE
Repeated ACR Pull for about 5 minutes after kubectl apply

### ISSUE THREAD

**Issue by nmiodice** [Azure/aad-pod-identity#684]
**Describe the bug**
I am seeing a behavior that I'm not 100% sure is related to `aad-pod-identity`. When new pods spin up, I see a behavior where they are unable to pull images from ACR for about 5 - 10 minutes, after which there are no issues pulling.

I can see in the `mic` controller logs that the identity is assigned quickly to the pod, and the identity that the pod runs under has the requisite `AcrPull` permissions for the cluster. The cluster is provisioned using managed identity, and the kubelet identity also has this RBAC applied.

**Steps To Reproduce**

[partial] Sample Terraform:

```hcl
# https://www.terraform.io/docs/providers/azurerm/r/user_assigned_identity.html
resource "azurerm_user_assigned_identity" "pi_common" {
  resource_group_name = data.azurerm_kubernetes_cluster.kube.node_resource_group
  location            = data.azurerm_resource_group.rg.location
  name                = "..."
  tags                = local.tags
}

# https://www.terraform.io/docs/providers/azurerm/r/user_assigned_identity.html
resource "azurerm_user_assigned_identity" "pi_exports" {
  resource_group_name = data.azurerm_kubernetes_cluster.kube.node_resource_group
  location            = data.azurerm_resource_group.rg.location
  name                = "..."
  tags                = local.tags
}

locals {
  managed_identities = [
    azurerm_user_assigned_identity.pi_common.principal_id,
    azurerm_user_assigned_identity.pi_exports.principal_id,
  ]
}

# https://www.terraform.io/docs/providers/azurerm/r/role_assignment.html
resource "azurerm_role_assignment" "acrpull_mi" {
  count                = length(local.managed_identities)
  scope                = var.acr_id
  role_definition_name = "AcrPull"
  principal_id         = local.managed_identities[count.index]
}

# https://www.terraform.io/docs/providers/azurerm/r/role_assignment.html
resource "azurerm_role_assignment" "acrpull" {
  scope                = var.acr_id
  role_definition_name = "AcrPull"
  principal_id         = data.azurerm_kubernetes_cluster.kube.kubelet_identity[0].object_id
}

# https://www.terraform.io/docs/providers/azurerm/d/resource_group.html
data "azurerm_resource_group" "kube_rg" {
  name = data.azurerm_kubernetes_cluster.kube.node_resource_group
}

# https://www.terraform.io/docs/providers/azurerm/r/role_assignment.html
resource "azurerm_role_assignment" "mi_operator" {
  scope                = data.azurerm_resource_group.kube_rg.id
  role_definition_name = "Managed Identity Operator"
  principal_id         = data.azurerm_kubernetes_cluster.kube.kubelet_identity[0].object_id
}

resource "azurerm_role_assignment" "vm_contrib" {
  scope                = data.azurerm_resource_group.kube_rg.id
  role_definition_name = "Virtual Machine Contributor"
  principal_id         = data.azurerm_kubernetes_cluster.kube.kubelet_identity[0].object_id
}

```

**Expected behavior**

The ACR Pull should work as soon as the pod comes up

**AAD Pod Identity version**

latest

**Kubernetes version**
1.15

**Additional context**

Sample error logs:

```
  Warning  Failed     20s                kubelet, aks-default-37377463-vmss000002  Failed to pull image "<MASKED>:latest": rpc error: code = Unknown desc = Error response from daemon: Get https://<MASKED>/latest: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
```

**Comment by chewong** [comment]
Are you using managed identity to pull the image? Relevant issue: https://github.com/kubernetes/kubernetes/issues/92326

**Comment by nmiodice** [comment]
We are. I'm not actually sure whether or not the pod identity or the kubelet identity will be used for the pull. Do you know of any workaround, or when/if that change will be rolled out for GA?

**Comment by aramase** [comment]
@nmiodice `kubelet` identity is used for image pull. All the control plane components (kubelet, controller manager) run on the host and are not intercepted by pod-identity. The PR for that issue is open and is being reviewed. I believe once it merges, it'll land in the next release. You should be able to check on the issue if there are currently any workarounds.

**Comment by nmiodice** [comment]
Ok thanks! I'll close this out, as the issue is unrelated to aad-pod-identity

**Comment by nmiodice** [comment]
For any curious readers... This issue has nothing to do with `aad-pod-identity` but I've been asked a few times (and referred back to this GH issue) so I'll post this here:

**Root Cause**
The issue surfaced because of a credential caching issue in the underlying kubernetes codebase. At a high level, the credentials for a Managed Identity ACR would be wiped out from the cache in the case that an image pull was required for a non Managed Identity container registry. I don't know the specifics of this but the issue gets worse as you add more container registry sources to your deployment.
 
In our scenario, we were pulling application images from ACR. We had periodic cron jobs that needed a curl base image that was pulled from a public container registry. Each time the job started (about every 5 minutes), the image would pull and evict any MI credentials from the cache. Annoying...
 
**Workaround**
The workaround we leveraged was to build (or import) any external container images into a single ACR so that AKS would only ever need to pull from one ACR. This circumvented the cache invalidation issue because most of the time, we only ever pull from a single container registry. We did this with our curl image and it basically fixed our issue.
 
Some components like `aad-pod-identity` still pull from a public container registry. These pull so infrequently that the cache invalidation issue is not really a problem for us.
